### PR TITLE
Implementation, Refact

### PR DESCRIPTION
**Pull request by goulashify**
- [x] Renamed template parameters to 1 character long, added comment
- [x] Renamed Adaptor to Transformer
- [x] Removed transformer from CommonRepository, it returns entities now
- [x] Added `CommonService<E, D>`
- [x] Removed DTO's between service and persistence layers

### CODE REVIEW

File: src/main/java/com/danigu/blog/comment/persistence/CommentRepository.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * @author dani
+ */
+public class CommentRepository extends CommonRepository<CommentEntity> {
+    public CommentRepository(EntityManagerFactory emf) {
+        super(emf);
+    }
+
+    public List<CommentEntity> getAllWithPostId(long id) {
+        TypedQuery<CommentEntity> query = getEntityManager()
+                .createQuery("SELECT c from Comment c WHERE c.post.id = :postId", getClazz());
```
Comment (goulashify) [inline src/main/java/com/danigu/blog/comment/persistence/CommentRepository.java]: Ilyet lehet csinalni generikusokkal?
A `CommonRepository`-ban van ilyen a generikus lekeresre:

```java
EntityManager em = getEntityManager();

CriteriaQuery<E> cq = em.getCriteriaBuilder().createQuery(getClazz());
cq.select(cq.from(getClazz()));

return em.createQuery(cq).getResultList();
```

Nekem sokkal jobban tetszik ez a "string" query dolog + egy juniornak erthetobb lehet ez, van valami megoldas arra hogy hogyan tudnam hasonloan leirni a `getAll()`-t (fent) a `CommonRepository`-ban (ugy hogy csak class reference-em van az Entity-rol)?

File: src/main/java/com/danigu/blog/comment/service/CommentService.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+         * persisting any Post, since it's not guaranteed that {@link Post} can only be created by the PostService
+         * we might be persisting Post's here (as a cascade) which is out of scope.
+         *
+         * This would be avoidable by depending on an instance of {@link com.danigu.blog.post.service.PostService},
+         * but then we're tightly coupled to that. :(
+         */
+        Comment dto = new Comment(Long.valueOf(null), post, content);
+        CommentEntity entity = transformer.convertFrom(dto);
+
+        return transformer.convert(repository.save(entity));
```
Comment (goulashify) [inline src/main/java/com/danigu/blog/comment/service/CommentService.java]: Itt az a bajom hogy valahogyan validalni kellene azt hogy a `Post` letezik, viszont ha validalni akarjuk akkor coupled lesz a `CommentService` a `PostService`-hez, van valami megoldas amivel ezt ugyesen lehetne?

Usually service-service dependency volt ahogyan eddig csinaltam, van erre valami okosabb megoldas amivel talan megtarthatjuk a loose couplingot a 2 service kozott vagy tok normal es elfogadott a service-level coupling?

File: src/main/java/com/danigu/blog/comment/service/CommentService.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    public List<Comment> getComments(Post post) {
+        checkNotNull(post);
```
Comment (goulashify) [inline src/main/java/com/danigu/blog/comment/service/CommentService.java]: Itt tulloves a `RuntimeException` nem?
`IllegalArgumentException`-t viszont burden adni minden ilyen kereshez a kliensoldalrol, szerinted jo apporach az ha csak egy ures `List<Comment>`-et adok vissza ha null a `post` argument?

File: src/main/java/com/danigu/blog/comment/service/CommentService.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+public class CommentService extends CommonService<CommentEntity, Comment> {
+    public CommentService(CommentRepository repository, CommentEntityDTOTransformer transformer) {
+        super(repository, transformer);
+    }
+
+    protected Class<CommentEntity> getEntityClazz() {
+        return CommentEntity.class;
+    }
+
+    public Comment comment(Post post, String content) {
```
Comment (manokovacs) [inline src/main/java/com/danigu/blog/comment/service/CommentService.java]: I would rather name this method as what it does ("insert" or "add").

File: src/main/java/com/danigu/blog/comment/Comment.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+import com.danigu.blog.post.Post;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+/**
+ * @author dani
+ */
+@AllArgsConstructor
+@Getter
```
Comment (manokovacs) [inline src/main/java/com/danigu/blog/comment/Comment.java]: Recommend @Data annotation instead.

File: src/main/java/com/danigu/blog/comment/persistence/CommentEntity.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package com.danigu.blog.comment.persistence;
```
Comment (manokovacs) [inline src/main/java/com/danigu/blog/comment/persistence/CommentEntity.java]: I would consider CommentEntity as part of the domain model, not part of the persistence layer. If you are using ORM that maps domain objects into the database, your persistence layer would only map business objects into the db.

File: src/main/java/com/danigu/blog/common/TwoWayTransformer.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package com.danigu.blog.common;
+
+/**
+ * Interface for two-way transformer.
+ * @param <F> from
+ * @param <T> to
+ */
+public interface TwoWayTransformer<F, T> extends OneWayTransformer<F, T> {
```
Comment (manokovacs) [inline src/main/java/com/danigu/blog/common/TwoWayTransformer.java]: I think using the simply "Transformer" is sufficient. I would recommend to add fromEntity and toEntity methods.

**Review by goulashify — COMMENTED** [review]
Added some questions.

File: src/main/java/com/danigu/blog/comment/persistence/CommentEntity.java (hunk 6 above)
Comment (goulashify) [reply src/main/java/com/danigu/blog/comment/persistence/CommentEntity.java]: Moved to the package of the domain.

File: src/main/java/com/danigu/blog/common/TwoWayTransformer.java (hunk 7 above)
Comment (goulashify) [reply src/main/java/com/danigu/blog/common/TwoWayTransformer.java]: Renamed to `Transformer<E, D>` with the suggested methods.

File: src/main/java/com/danigu/blog/comment/persistence/CommentRepository.java (hunk 1 above)
Comment (manokovacs) [reply src/main/java/com/danigu/blog/comment/persistence/CommentRepository.java]: Arbitrary. I would only recommend not to expose the internal, persistence specific implementation details to the interface of the Repository. Meaning, I would avoid having criteria object as parameter of public repository interface. You may implement protected method in the CommonRepository like that, though.

File: src/main/java/com/danigu/blog/comment/service/CommentEntityDTOTransformer.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+import com.danigu.blog.common.TwoWayTransformer;
+import com.danigu.blog.post.persistence.PostEntity;
+import com.danigu.blog.post.service.PostEntityDTOTransformer;
+import lombok.AllArgsConstructor;
+
+/**
+ * @author dani
+ */
+@AllArgsConstructor
+public class CommentEntityDTOTransformer implements TwoWayTransformer<CommentEntity, Comment> {
```
Comment (manokovacs) [inline src/main/java/com/danigu/blog/comment/service/CommentEntityDTOTransformer.java]: I would call this simply CommentTransformer.

File: src/main/java/com/danigu/blog/common/service/CommonService.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    public void deleteById(Long id) throws NotFoundException {
+        if(isNull(id)) throw new NotFoundException("No " + getEntityClazz().getName() + " with id 'null'.");
+
+        E entity = repository.getById(id);
+        if(isNull(entity)) throw new NotFoundException(getEntityClazz().getName() + " not found.");
+
+        repository.remove(entity);
+    }
+
+    protected abstract Class<E> getEntityClazz();
```
Comment (manokovacs) [inline src/main/java/com/danigu/blog/common/service/CommonService.java]: I would recommend having this as a required constructor argument.

File: src/main/java/com/danigu/blog/comment/service/CommentService.java (hunk 2 above)
Comment (manokovacs) [reply src/main/java/com/danigu/blog/comment/service/CommentService.java]: Post-comment relation is composition, meaning the comments do not exist without the post. You could even add the `getPostComments()` method to the PostService.

File: src/main/java/com/danigu/blog/comment/service/CommentService.java (hunk 3 above)
Comment (manokovacs) [reply src/main/java/com/danigu/blog/comment/service/CommentService.java]: If the `post` is empty, do not return empty list, but throw an exception. RuntimeException is fine.

File: src/main/java/com/danigu/blog/common/service/CommonService.java (hunk 9 above)
Comment (goulashify) [reply src/main/java/com/danigu/blog/common/service/CommonService.java]: Refactored, it's in.

File: src/main/java/com/danigu/blog/comment/service/CommentEntityDTOTransformer.java (hunk 8 above)
Comment (goulashify) [reply src/main/java/com/danigu/blog/comment/service/CommentEntityDTOTransformer.java]: Refactored as well.

File: src/main/java/com/danigu/blog/comment/service/CommentService.java (hunk 3 above)
Comment (goulashify) [reply src/main/java/com/danigu/blog/comment/service/CommentService.java]: Ok, `checkNotNull` does that!

**Pull request merged**